pkg/proxy/userspace/proxier.go

```diff
@@ -68,6 +68,13 @@ type ServiceInfo struct {
 	stickyMaxAgeSeconds int
 	// Deprecated, but required for back-compat (including e2e)
 	externalIPs []string
+
+	// started is incremented when the service is created, and decremented
+	// when the service's socket beings accepting requests
+	started *sync.WaitGroup
+	// finished is incremented when the service is created, and decremented
+	// when the service's socket shuts down
+	finished *sync.WaitGroup
```

Comment:
How heavyweight are WaitGroups? Given that this is only used for testing, not for ordinary functioning of the proxy.

Reply:
I had the same thought

Reply:
WaitGroup just uses atomic ops. Not significantly heavier weight IMO than the previous atomic ops stuff.

Reply:
https://golang.org/src/sync/waitgroup.go?s=2022:2057#L43 looks a lot heavier, especially for something we are only using in a unit test

Reply:
This comment is still outstanding

Reply:
@liggitt We don't actually need waitgroup values > 1, so if you'd like though, I'm happy to move it to custom-rolled atomics and a busyloop for wait.

But I would disagree that watigroup add/done is heavy especially since there's no contention whatsoever when testcases are not running.

Reply:
for something used only by tests, I'd like the cost to be as close to zero for the production code as possible

Reply:
changed to use atomics for started/finished rather than waitgroups.